### PR TITLE
Lsp Spec

### PR DESCRIPTION
**Pull request by niteshbalusu11**
Signed-off-by: Nitesh Balusu <84944042+niteshbalusu11@users.noreply.github.com>

This is an implementation of LSPS0 and LSPS1 for both client and server.

https://github.com/BitcoinAndLightningLayerSpecs/lsp/blob/main/LSPS1/README.md

I did not add code comments yet expecting changes, please review the code and let me know and I can add comments at the end.

### CODE REVIEW

**Comment by alexbosworth** [comment]
Instead of request_info / buy_channel / check_order for `bos get-inbound-channel`

I think it could be `--dryrun` or `--recovery <orderId>` if you want to only request info or only check an existing flow. This would be consistent with the other commands where you interact with a service and it would leave the command execution itself with the possibility of just working with minimal input

`--amount` could default to 5mm, that is consistent with `open`
`--priority` could be `--max-expected-wait-hours` and then convert hours to blocks like hours * 6, default to 48 hours. Blocks isn't possible to guarantee, especially at the low end, so let's not give this impression
`--pubkey` can be the argument: so `bos get-inbound-channel 023049203498` - this signals that it's a required option

`bos offer-channel-open` seems good if it doesn't require any arguments

`--fee-rate` could be `--capacity-fee-rate` to distinguish it from the channel fee rate, otherwise it may be confusing
`--max-capacity`, `--min-capacity` should be a number, can change to string for the wire protocol
`--min-onchain-payment-size` I don't see why this is necessary, can omit it

One thing I'm thinking about is like if you can make something in memory when there is a request, is that a DoS vector? Instead maybe it could just not register anything in memory until payment is apparent? Didn't go super deep on that to confirm what's happening there though

**Comment by niteshbalusu11** [comment]
Will make the changes.
It only registers in memory when the client asks for an invoice the first time because the spec asks for specific states to be present when the client submits for an order and the server returns the invoice/address.

File: lsp/buy_channel.js [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+          }
+        } catch (err) {
+          return cbk([400, 'ExpectedValidPaymentRequestToBuyChannel', {err}]);
+        }
+
+        return cbk();
+      }],
+
+      // Ask to buy the channel
+      ask: ['validateMessage', ({}, cbk) => {
```
Comment (alexbosworth) [inline lsp/buy_channel.js:126]: should probe first to see what the routing fee is going to cost and then feed that routing fee back into the pay call

File: lsp/lsps1_client.js [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+        
+        if (!knownTypes.includes(args.type)) {
+          return cbk([400, 'ExpectedKnownChannelTypeToRunLspClient']);
+        }
+        
+        return cbk();
+      },
+      
+      // Get node info
+      getNodeInfo: ['validate', ({}, cbk) => {
```
Comment (alexbosworth) [inline lsp/lsps1_client.js:69]: could use `connectPeer` in ln-sync here instead

File: lsp/lsps1_client.js [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+              args.logger.info({is_payment_sent: true});
+            }
+          } catch (err) {
+            args.logger.error({err});
+          }
+        });
+        
+        sub.on('error', err => {
+          sub.removeAllListeners();
```
Comment (alexbosworth) [inline lsp/lsps1_client.js:155]: should probably log the error here

File: lsp/process_order.js [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+          orderResponse.result.payment.state = constants.paymentStates.expectPayment;
+          orderResponse.result.payment.fee_total_sat = String(fees);
+  
+          // Same as fees because push amounts are not supported
+          orderResponse.result.payment.order_total_sat = String(fees);
+
+          const secret = randomSecret();
+          const invoiceId = sha256(secret);
+  
+          const {request, id} = await createHodlInvoice({expires_at: expiresAt, id: invoiceId, lnd: args.lnd, tokens: fees});
```
Comment (alexbosworth) [inline lsp/process_order.js:300]: you don't need to make a secret and invoiceId, the secret can be generated by the method itself

File: lsp/process_order.js [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+
+        const timeout = setTimeout(() => {
+          sub.removeAllListeners();
+
+          return cbk([503, 'TimedOutWaitingForLightningPayment']);
+        },
+        orderExpiryMs);
+
+        sub.on('invoice_updated', async invoice => {
+          try {
```
Comment (alexbosworth) [inline lsp/process_order.js:348]: since there are a bunch of async things that happen here, this could be abstracted out to its own file

**Comment by niteshbalusu11** [comment]
All changes done.

**Comment by alexbosworth** [comment]
I'll take a look at this over the weekend please don't make changes

**Comment by alexbosworth** [comment]
i merged it but it couldn't do it cleanly because of rebase, you can take a look at the changes though to see if everything looks ok

**Pull request closed without merging**